### PR TITLE
[JEP-305] Incrementals profile

### PR DESCRIPTION
**Pull request by jglick**
[JEP-305](https://github.com/jenkinsci/jep/blob/master/jep/305/README.adoc)

An example so far: https://github.com/jenkinsci/git-plugin/pull/582 consumed by https://github.com/jenkinsci/structs-plugin/pull/36, consumed in turn by https://github.com/jenkinsci/junit-plugin/pull/102 along with https://github.com/jenkinsci/jenkins/pull/3394.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -817,6 +815,7 @@
       <id>jenkins-release</id>
       <properties>
         <skipTests>${release.skipTests}</skipTests>
+        <findbugs.skip>${release.skipTests}</findbugs.skip>
```
Comment (oleg-nenashev) [inline pom.xml]: unrelated but LGTM

Comment (jglick) [reply pom.xml]: Yeah could be filed separately; just noticed that deployment was unnecessarily slow.

File: pom.xml [hunk 2]
```diff
@@ -484,15 +486,17 @@
...
                 <requireMavenVersion>
-                  <version>[3.1.0,)</version>
-                  <message>3.1.0 required by frontend-maven-plugin at least.</message>
+                  <version>[3.3.1,)</version>
+                  <message>3.3.1 required to at least look at .mvn/ if it exists.</message>
                 </requireMavenVersion>
                 <requireJavaVersion>
                   <version>[1.8.0,)</version>
                 </requireJavaVersion>
+                <!-- TODO failing during incrementals deploy: MENFORCER-281
```
[the remark this answers is not in the document]
Comment (jglick) [reply pom.xml]: No 3.0.0-M1 was the _affected_ version; the fix https://github.com/apache/maven-enforcer/commit/ad0a84368f97a335a9455f04ec04bfbbb10ac49e is targeted to 3.0.0.

**Comment by jglick** [comment]
> The proposed patch disables enforcer's snapshot checks everywhere for `release` profile, not just for incremental releases.

Yes, see comment above—MRP does this anyway AFAIK.

> As a MENFORCER-281 workaround, you could create a profile which adds `requirePluginVersions` only when there is no incrementals deployment

Or only in the `jenkins-release` profile. Can try it.

**Comment by jglick** [comment]
> Can try it.

Nope, now I remember why I did not do this to begin with—it will not work. See [MENFORCER-281](https://issues.apache.org/jira/browse/MENFORCER-281). If `flatten-maven-plugin` is in use, which it would be at all times for projects potentially producing incrementals, then you cannot use this rule until 3.0.0 is released.

File: pom.xml [hunk 3]
```diff
@@ -542,10 +546,6 @@
                     <exclude>commons-logging:commons-logging:*:jar:runtime</exclude>
                   </excludes>
                 </bannedDependencies>
-                <requireReleaseDeps>
-                  <message>No Snapshots Allowed For Release Versions</message>
-                  <onlyWhenRelease>true</onlyWhenRelease>
```
[the remark this answers is not in the document]
Comment (jglick) [reply pom.xml]: https://github.com/jglick/mrp-test/commit/3e3a0ecd5aa1081a9a701e3518c5c54d8681c1a9 confirms that _this_ rule was gratuitous:

```
org.apache.maven.shared.release.ReleaseFailureException: Can't release project due to non released dependencies :
    test:simple-maven-project-with-tests:jar:1.0-SNAPSHOT:compile
in project 'two' (test:two:jar:2.40-SNAPSHOT)
    at org.apache.maven.shared.release.phase.CheckDependencySnapshotsPhase.checkProject (CheckDependencySnapshotsPhase.java:182)
    at org.apache.maven.shared.release.phase.CheckDependencySnapshotsPhase.execute (CheckDependencySnapshotsPhase.java:108)
    at org.apache.maven.shared.release.DefaultReleaseManager.prepare (DefaultReleaseManager.java:234)
    at org.apache.maven.shared.release.DefaultReleaseManager.prepare (DefaultReleaseManager.java:169)
    at org.apache.maven.shared.release.DefaultReleaseManager.prepare (DefaultReleaseManager.java:146)
    at org.apache.maven.shared.release.DefaultReleaseManager.prepare (DefaultReleaseManager.java:107)
    at org.apache.maven.plugins.release.PrepareReleaseMojo.prepareRelease (PrepareReleaseMojo.java:286)
    at org.apache.maven.plugins.release.PrepareReleaseMojo.execute (PrepareReleaseMojo.java:240)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:208)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:154)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:146)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:290)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:194)
```

**Comment by jglick** [comment]
> Maybe doing a custom build of a plugin would be preferable

I would definitely consider this if no official release seems imminent. It looks like we have done so for this very plugin in the past.

**Comment by basil** [comment]
Causes https://github.com/jenkins-infra/backend-extension-indexer/issues/58 — any plugin that has been incrementalified uses `flatten-maven-plugin`, which removes the Jenkins core dependency and breaks the logic in `jenkins-infra/backend-extension-indexer`, which relies on the Jenkins core dependency being on the classpath in order to compile the plugin and list its extensions. Suggest updating the `flatten-maven-plugin` configuration to retain the Jenkins core dependency or updating the logic in `jenkins-infra/backend-extension-indexer` to re-add the Jenkins core dependency to the classpath, if not rewriting `jenkins-infra/backend-extension-indexer` from scratch.

**Pull request merged**